libraries/chain/fork_database.cpp

```diff
@@ -354,18 +354,19 @@ namespace eosio::chain {
          EOS_RETHROW_EXCEPTIONS( fork_database_exception, "serialized fork database is incompatible with configured protocol features" )
       }
 
-      if (mark_valid == mark_valid_t::yes)
-         bs_accessor_t::set_valid(*n, true);
-
       auto inserted = index.insert(n);
-      if( !inserted.second ) {
-         if( ignore_duplicate == ignore_duplicate_t::yes ) return;
-         EOS_THROW( fork_database_exception, "duplicate block added", ("id", n->id()) );
+      if( !inserted.second && ignore_duplicate != ignore_duplicate_t::yes ) {
```

Comment:
Now we will continue processing `if ignore_duplicate == ignore_duplicate_t::yes`?

Reply:
Yes, additional processing is needed for `mark_valid` case. Also `head` being updated is an expected side-affect of a call to `add`, so best to see if `head` needs to be updated.